express-notes-api/index.js

```diff
@@ -0,0 +1,91 @@
+const data = require('./data.json');
+const express = require('express');
+const fs = require('fs');
+const app = express();
+
+app.listen(3000, () => {
+});
+
+app.get('/api/notes', (req, res) => {
+  const notes = [];
+  for (const key in data.notes) {
+    notes.push(data.notes[key]);
+  }
+  res.json(notes);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  let note = {};
+  const id = req.params.id;
+  if (req.params.id < 1) {
+    res.status(400).json({ error: 'id must be a positive integer' });
+  } else if (req.params.id) {
+    note = data.notes[req.params.id];
+    if (note === undefined) {
+      res.status(404).json({ error: 'cannot find note with id ' + id });
+    }
+  }
+  res.json(note);
+});
+
+app.use(express.json());
+
+app.post('/api/notes', (req, res) => {
+  if (!req.body.content) {
+    res.status(400).json({ error: 'content is required field' });
+  } else if (req.body.content) {
+    const newNote = {};
+    newNote.id = data.nextId;
+    newNote.content = req.body.content;
+    data.notes[data.nextId] = newNote;
+    data.nextId++;
+    fs.writeFile('data.json', JSON.stringify(data, null, 2), err => {
+      if (err) {
+        res.status(500).json({ error: 'An unexpected error occurred.' });
+
+      }
+    });
+    res.status(201).json(newNote);
```

Comment:
You'll need to move this success response into the writeFile callback. We don't know if the note was actually saved until that callback's code block runs.